ui: Ameliorate TUI resize bug.

justinmk opened the pull request:
It's not ideal but seems to work quite well. Only enabled for the "built-in UI" (TUI) since we haven't seen reports of the resize issue for GUIs.

Fixing the root issue is still a priority but it will take time and there are more important things to do in the near term.

References #3929 #4884 #5692 #6157

brcolow commented on src/nvim/main.c:
```diff
@@ -429,6 +429,7 @@ int main(int argc, char **argv)
     // Stop reading from input stream, the UI layer will take over now.
     input_stop();
     ui_builtin_start();
+    do_cmdline_cmd("au VimResized * call timer_start(100, {-> execute('mode')})");
```
+1 for lambda

justinmk replied:
@brcolow Thanks for working on it :)

JelteF commented:
For me this doesn't fix the resize issue

justinmk commented:
@JelteF Could you try this variations (just enter them in the nvim command line, then try to resize):

    au VimResized * call timer_start(300, {-> execute('mode')})

    au VimResized * call timer_start(300, {-> execute('refresh!|mode')})

JelteF commented:
@justinmk The first one still doesn't work and the second one gives an error:
```
Error detected while processing function <lambda>8:                                                                   
line    1:                                                                                                            
E492: Not an editor command: refresh!|mode      
```

However, I tried to increase the timer of the first one a bit more, to 1000, and that worked great. You can still have a short visual glitch, but then it corrects it.

justinmk commented:
Oops it should be redraw. Ok, thanks

justinmk commented:
Does 500 work?

tweekmonster commented:
Wouldn't you want to cancel any previous timers?  Fast resizing might leave us with `:mode` spam making the screen flicker like the timer was never used.

JelteF commented:
Nope, but 800 works for both. One thing I noticed was that when you resize while typing a `:` command (ex-mode?), the screen is fully cleared except for the line with the `:` command.

@tweekmonster that sounds reasonable.

justinmk commented:
The root cause has to do with the input queue timing. I looked at it in November, didn't go further yet.

choco commented:
Tested the fix above and even though it works when splitting the tmux pane, when I destroy the same split pane neovim doesn't resize itself most of the times. Changing the timing doesn't seem to change this from happening, but lower timing influences the first part of the issue (when splitting)

tweekmonster committed:
ui: Fix TUI resize bug

statusline still disappears in some cases, but this change is a net
improvement.

References #3929 #5692 #4884 #6157

Tranquility commented:
I also realized that :redraw does not help. I have the problem every time a destroy a horizontal tmux pane. But the resize works correctly if a use a vertical tmux pane. Therefore the only way to redraw and for me right now is creating and directly destroying a vertical tmux pane.

oni-link commented on src/nvim/ui.c:
```diff
@@ -171,6 +171,11 @@ void ui_refresh(void)
+  if (updating_screen) {
+    loop_schedule(&main_loop, event_create(1, ui_refresh_event, 0));
+    return;
+  }
```
`if` block could be moved before the declarations and `loop_schedule(...)` could be replaced with `ui_schedule_refresh()`.

tweekmonster replied:
> if block could be moved before the declarations

Good point.  The current placement was due to me trying different things.  If it gets here, it usually happens a dozen or so times.  So, I was testing the width and height to double check that it wasn't from multiple SIGWINCH events.

> and loop_schedule(...) could be replaced with ui_schedule_refresh().

![](http://sae.tweek.us/media/emoticons/emot-doh.gif)

justinmk commented:
Any airline users seeing an improvement from this?

choco commented:
This doesn't solve the https://github.com/neovim/neovim/issues/4884 issue for me, with or without airline. But at least it works sometimes ( 1 out of 10 times? ), before it never worked.

tweekmonster commented:
@justinmk I didn't notice anything different in regard to airline.

@choco Are you using the most recent changes from this PR?  It sounds like you're describing the original PR using timers.

I used to have the problem mentioned in that issue, but I can no longer reproduce it.  This PR makes it impossible for a resize to be ignored, so if you're still having issues it must mean that nvim isn't noticing the terminal is resizing.

justinmk committed:
lint

You cannot escape clint...

choco commented:
Yep, still having it unfortunately :( But I noticed that removing `lazyredraw` from my vimrc seems to fix the issue after this change. Is there any way to ignore the lazyredraw value in these resizing situations?

JelteF commented:
This seems to fix the issue for me.
